source/server/listener_manager_impl.cc

```diff
@@ -490,36 +521,48 @@ ProtobufTypes::MessagePtr ListenerManagerImpl::dumpListenerConfigs() {
     // To avoid confusion in config dump, in that case, we add these listeners to warming
     // listeners config dump rather than active ones.
+    DynamicListener* dynamic_listener =
+        getOrCreateDynamicListener(listener->name(), *config_dump, listener_map);
+
+    DynamicListenerState* dump_listener;
     if (workers_started_) {
-      dump_listener = config_dump->mutable_dynamic_active_listeners()->Add();
+      dump_listener = dynamic_listener->mutable_active_state();
     } else {
-      dump_listener = config_dump->mutable_dynamic_warming_listeners()->Add();
+      dump_listener = dynamic_listener->mutable_warming_state();
     }
-    dump_listener->set_version_info(listener->versionInfo());
-    dump_listener->mutable_listener()->MergeFrom(listener->config());
-    TimestampUtil::systemClockToTimestamp(listener->last_updated_,
-                                          *(dump_listener->mutable_last_updated()));
+    fillState(*dump_listener, *listener);
   }
 
   for (const auto& listener : warming_listeners_) {
-    auto& dynamic_listener = *config_dump->mutable_dynamic_warming_listeners()->Add();
-    dynamic_listener.set_version_info(listener->versionInfo());
-    dynamic_listener.mutable_listener()->MergeFrom(listener->config());
-    TimestampUtil::systemClockToTimestamp(listener->last_updated_,
-                                          *(dynamic_listener.mutable_last_updated()));
+    DynamicListener* dynamic_listener =
+        getOrCreateDynamicListener(listener->name(), *config_dump, listener_map);
+    DynamicListenerState* dump_listener = dynamic_listener->mutable_warming_state();
+    fillState(*dump_listener, *listener);
   }
 
-  for (const auto& listener : draining_listeners_) {
-    auto& dynamic_listener = *config_dump->mutable_dynamic_draining_listeners()->Add();
-    dynamic_listener.set_version_info(listener.listener_->versionInfo());
-    dynamic_listener.mutable_listener()->MergeFrom(listener.listener_->config());
-    TimestampUtil::systemClockToTimestamp(listener.listener_->last_updated_,
-                                          *(dynamic_listener.mutable_last_updated()));
+  for (const auto& draining_listener : draining_listeners_) {
+    const auto& listener = draining_listener.listener_;
+    DynamicListener* dynamic_listener =
+        getOrCreateDynamicListener(listener->name(), *config_dump, listener_map);
+    DynamicListenerState* dump_listener = dynamic_listener->mutable_draining_state();
+    fillState(*dump_listener, *listener);
+  }
+
+  for (const auto& state_and_name : error_state_tracker_) {
+    DynamicListener* dynamic_listener =
+        getOrCreateDynamicListener(state_and_name.first, *config_dump, listener_map);
+
+    const envoy::admin::v2alpha::UpdateFailureState& state = *state_and_name.second;
+    dynamic_listener->mutable_error_state()->CopyFrom(state);
+  }
+
+  // Dump errors not associated with named listeners.
```

Comment:
Why not have an explicit field for this, rather than an unnamed listener. The latter seems a bit hacky and harder to parse out.

Reply:
Because each of these is a failed listener update, it just failed before we had a name.

I think you're seeing this as fundamentally different (we should aggregate all the failure and dump them together) where I see it as the internals of where we create the listener in the parsing process,and each failure message is functionally a listener creation failure.  I don't think we want to batch them together, I think they should each be tracked separately as individual failures.